src/zkevm_specs/evm/execution/log.py

```diff
@@ -0,0 +1,55 @@
+from ..instruction import Instruction, Transition
+from ..table import CallContextFieldTag, TxLogFieldTag, TxContextFieldTag
+from ..opcode import Opcode
+from ..execution_state import ExecutionState
+from ...util.param import LOG_STATIC_GAS
+
+
+def log(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    # constrain op in [log0, log4] range
+    instruction.range_lookup(opcode - Opcode.LOG0, 5)
+
+    # pop `mstart`, `msize` from stack
+    mstart = instruction.rlc_to_fq_exact(instruction.stack_pop(), 8)
+    msize = instruction.rlc_to_fq_exact(instruction.stack_pop(), 8)
+
+    # check contract_address in CallContext & TxLog
+    # use call context's  callee address as contract address
+    contract_address = instruction.call_context_lookup(CallContextFieldTag.CalleeAddress)
+    instruction.constrain_equal(contract_address, instruction.tx_log_lookup(TxLogFieldTag.Address))
+    # check not static call
+    instruction.constrain_equal(0, instruction.call_context_lookup(CallContextFieldTag.IsStatic))
+
+    # constrain topics in stack & logs
+    for i in range(int(opcode) - Opcode.LOG0):
```

Comment:
you should expand this part. The current for loop didn't reflect the logic in the circuit code

Reply:
updated in [69f93be](https://github.com/appliedzkp/zkevm-specs/pull/125/commits/69f93be980b266701a44d3070e0fc8000dceadb5) thanks